### PR TITLE
Fix connection leak when client disconnected

### PR DESCRIPTION
**Pull request by sorah**
established connections were leaked when client disconnected before proxyreq
completes. This happens all time for wait=true requests.

### CODE REVIEW

**Comment by yichengq** [comment]
this helps #2743

**Comment by sorah** [comment]
I didn't know existing issue; this should help that

File: proxy/reverse.go [hunk 1]
```diff
@@ -98,7 +113,10 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 		return
 	}
 
-	defer res.Body.Close()
+	defer func() {
+		completeCh <- true
```
Comment (yichengq) [inline proxy/reverse.go]: could we move this defer early? if it returns when res == nil, the go routine will not be recycled.

**Comment by yichengq** [comment]
could we add package name to commit title `proxy: fix connection leak on wait when client disconnected`

**Comment by yichengq** [comment]
I have manually tested this PR, and it does fix the connection leak problem on orphan watch requests. Moreover, it doesn't affect proxy performance on local cluster testing.

**Comment by xiang90** [comment]
We need to separate the defer of the body close and complete. Or there might be a race as @yichengq described. LGTM.

**Comment by sorah** [comment]
Done.

File: proxy/reverse.go [hunk 2]
```diff
@@ -73,6 +73,25 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
...
+				if ok {
+					tp.CancelRequest(proxyreq)
+				}
+			case <-completeCh:
+			}
+		}()
+
+		defer func() {
+			completeCh <- true
+		}()
```
Comment (yichengq) [inline proxy/reverse.go]: one line?

Comment (sorah) [inline proxy/reverse.go]: I thought too and tried, but it's not valid on go 1.4.

Comment (yichengq) [inline proxy/reverse.go]: k

**Comment by yichengq** [comment]
lgtm

**Comment by yichengq** [comment]
Thanks for contribution!

**Comment by sorah** [comment]
:sparkles: you're welcome.

**Pull request merged**